src/libraries/Microsoft.PowerFx.Core/Entities/External/DataSourceInfo.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Entities.Delegation;
+using Microsoft.PowerFx.Core.Entities.QueryOptions;
+using Microsoft.PowerFx.Core.Functions.Delegation;
+using Microsoft.PowerFx.Core.Functions.Delegation.DelegationMetadata;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.UtilityDataStructures;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Core.Entities
+{
+    // Implements a base data source, used in DType.AssociatedDataSources, itself in RecordType constructor to host a CDP record type
+    internal class DataSourceInfo : IExternalTabularDataSource
+    {
+        public const string IsChoiceValue = "Value";
+
+        public readonly Dictionary<string, string> ColumnsWithRelationships;
+
+        public DataSourceInfo(AggregateType recordType, DisplayNameProvider displayNameProvider, TableDelegationInfo delegationInfo)
+        {
+            string GetDisplayName(string fieldName) => displayNameProvider == null || !displayNameProvider.TryGetDisplayName(new DName(fieldName), out DName displayName) ? fieldName : displayName.Value;
+
+            EntityName = new DName(delegationInfo.TableName);
+            IsWritable = !delegationInfo.IsReadOnly;
+            _delegationInfo = delegationInfo;
+            _type = recordType._type;
+            RecordType = (RecordType)recordType;
+
+            IEnumerable<string> fieldNames = displayNameProvider.LogicalToDisplayPairs.Select(pair => pair.Key.Value);
+
+            _displayNameMapping = new BidirectionalDictionary<string, string>(fieldNames.Select(f => new KeyValuePair<string, string>(f, GetDisplayName(f))).ToDictionary(kvp => kvp.Key, kvp => kvp.Value));
```

Comment:
Shouldn't this just be 
`new BidirectionalDictionary<string, string>(displayNameProvider.LogicalToDisplayPairs.ToDictionary(kvp => kvp.Key, kvp => kvp.Value));`